### DISCUSSION TITLE
invalidated forced dependencies

### DISCUSSION THREAD

**Discussion by tagteam** [ropensci/targets#1000] [General]
I ran into an error due to that the dependency on a target  is invalidated 
by assignment statements that modify the target, e.g.,
`
tar_deps({
   target1;
    target1$v = 1
})`
Here target1 is not recognized as a dependency.
I can fix this by copying the target:

`tar_deps({
target1;
x<- target1;
x$v=1})
`
If this is a feature would suggest that the wonderful documentation https://books.ropensci.org/targets/targets.html#dependencies mentions it.

**Comment by wlandau** [comment]
Perhaps the updated section at https://books.ropensci.org/targets/targets.html#dependencies can help. Dependency graph issues can arise in any number of ways, so I prefer to keep the advice in the manual as general as possible.

**Reply by tagteam** [reply]
yes, the example of the updated documentation is somewhat helpful. what I struggled to understand when targets returned the error was that any modification of a dependency, in the modified example below the global_object, within the target removes this dependency:
```tar_deps(
  function(argument) {
    local_object <- 1
    attr(global_object,"x") <- "1"
    argument + global_object + local_object + 2
  }
)
```
the result is 
```. + > [1] "{"      "+"      "<-"     "attr<-"```
i.e., no dependence on global_object detected, even though the code depends on global_object.
thank you for considering!

**Reply by wlandau** [reply]
`targets` uses `codetools::findGlobals()` in the backend to detect global dependencies, and that's where `tar_deps()` gets most of its behaviors

```r
codetools::findGlobals(
  function(argument) {
    local_object <- 1
    attr(global_object,"x") <- "1"
    argument + global_object + local_object + 2
  }
)
#> [1] "{"      "+"      "<-"     "attr<-"
```

`codetools` works most of the time, but like any [static code analysis system](https://bookdown.dongzhuoer.com/hadley/adv-r/ast-funs.html), it is not perfect. There are innumerable edge cases just like this one. In the general case, what I can say is that it is good practice to check `tar_visnetwork()` to verify the correct dependency relationships are detected.

**Reply by tagteam** [reply]
I see, it is out of your hands. Thanks a lot for the explanation, very helpful.
